#' Create phylogenetic tree
#' 
#' Create a phylogenetic tree from amino acid or nucleotide CDR3 sequences in a list of 
#' data frames.
#' 
#' @param list A list data frames of unproductive nucleotide sequences or productive 
#' nucleotide or amino acid sequenes generated by the LymphoSeq function productiveSeq 
#' where the parameter aggregate is set to "nucleotide".  vFamilyName, dFamilyName, and 
#' jFamilyName are required columns.
#' @param sample A character vector indicating the name of the sample in the productive
#' sequence list. 
#' @param type A character vector indicating whether "aminoAcid" or "nucleotide" sequences
#' should be aligned.  Available options are "nucleotide" and "aminoAcid".
#' @param method A character vector indicating the multiple sequence alignment method to 
#' be used.  Refer to the Bioconductor msa package for more details.  Options incude 
#' "ClustalW", "ClustalOmega", and "Muscle".
#' @param aa.model A character vector indicating the pairwise distance model to be used
#' to compare amino acid sequences.  Refer to the phangorn package for more details.
#' Available options incude "WAG", "JTT", "LG", "Dayhoff", "cpREV", "mtmam", "mtArt", "MtZoa",
#' "mtREV24", "VT","RtREV", "HIVw", "HIVb", "FLU", "Blossum62", "Dayhoff_DCMut" and "JTT_DCMut".
#' The default options is "JTT".
#' @param nt.model A character vector indicating the pairwise distance model to be used
#' to compare nucleotide sequences.  Refer to the phangorn package for more details.
#' Available options incude "JC69" or "F81".  The default options is "F81".
#' @param layout A character vector indicating the tree layout.  Options include 
#' "rectangular", "slanted", "fan", "circular", "radial" and "unrooted".
#' @return Returns a phylogenetic tree where each leaf represents a sequence color coded by the
#' V, D, and J gene usage.  The number next to each leaf refers to the sequence count.  A triangle 
#' shaped leaf indicates the dominant sequence.  Refer to the ggtree Bioconductor package 
#' documentation for details on how to manipulate the tree.
#' @examples
#' file.path <- system.file("extdata", "TCRB_sequencing", package = "LymphoSeq")
#' 
#' file.list <- readImmunoSeq(path = file.path)
#' 
#' productive.nt <- productiveSeq(file.list = file.list, aggregate = "nucleotide")
#' 
#' phyloTree(list = productive.nt, sample = "TCRB_Day1320_CD8_CMV", type = "nucleotide", 
#'          layout = "rectangular", method = "ClustalW", nt.model = "F81")
#' 
#' phyloTree(list = productive.nt, sample = "TCRB_Day1320_CD8_CMV", type = "aminoAcid", 
#'          layout = "circular", method = "ClustalOmega", nt.model = "JTT")
#' @export
#' @importFrom Biostrings DNAStringSet
#' @importFrom Biostrings AAStringSet
#' @importFrom phangorn as.phyDat
#' @importFrom phangorn dist.ml
#' @importFrom phangorn NJ
#' @import msa
#' @import ggtree
phyloTree <- function(list, sample, type = "nucleotide", method = "ClustalOmega", 
                      layout = "rectangular", aa.model = "JTT", nt.model = "F81") {
    file <- list[[sample]]
    if(length(grep(pattern = "vFamilyName|dFamilyName|jFamilyName", names(file))) != 3){
        stop("vFamilyName, dFamilyName, and jFamilyName are required columns.  Try aggregating the sequences by 'nucleotide' usng the function productiveSeq.", call. = FALSE)
    }
    if(nrow(file) < 3){
        stop("Cannot draw phlogenetic tree with less than 3 sequences.", call. = FALSE)
    }
    file[file == ""] <- "unresolved"
    if (type == "nucleotide") {
        file <- file[nchar(file$nucleotide) > 3, ]
        string <- Biostrings::DNAStringSet(file$nucleotide)
    }
    if (type == "aminoAcid") {
        file <- file[nchar(file$aminoAcid) > 3, ]
        string <- Biostrings::AAStringSet(file$aminoAcid)
    }
    names(string) <- file$count
    alignment <- msa::msa(string, method = method)
    if (type == "nucleotide") {
        alignment.phyDat <- phangorn::as.phyDat(x = alignment, type = "DNA", format = clustal)
        alignment.dist <- phangorn::dist.ml(alignment.phyDat, model = nt.model)
    }
    if (type == "aminoAcid") {
        alignment.phyDat <- phangorn::as.phyDat(x = alignment, type = "AA", format = clustal)
        alignment.dist <- phangorn::dist.ml(alignment.phyDat, model = aa.model)
    }
    tree <- phangorn::NJ(alignment.dist)
    geneFamilies <- paste(file$vFamilyName, file$dFamilyName, file$jFamilyName)
    geneFamilies <- gsub(geneFamilies, pattern = "IGH|IGL|IGK|TCRB|TCRA", replacement = "")
    geneFamilies <- gsub(geneFamilies, pattern = "unresolved", replacement = "UNR")
    tree.annotation <- data.frame(Names = file$count, Genomes = file$estimatedNumberGenomes, 
                                  Count = file$count, Genes = geneFamilies, 
                                  Frequency = round(file$frequencyCount, digits = 2), 
                                  Dominant = c("Yes", rep("No", length(string) - 1)))
    getPalette <- grDevices::colorRampPalette(RColorBrewer::brewer.pal(9, "Set1"))
    plot <- ggtree::ggtree(tree, layout = layout) %<+% tree.annotation + 
        ggtree::geom_tippoint(aes(color = Genes, shape = Dominant), size = 3) + 
        ggplot2::scale_color_manual(values = getPalette(length(unique(geneFamilies)))) + 
        ggplot2::guides(shape = FALSE) + 
        ggtree::geom_tiplab(size = 2, hjust = - 0.3) + 
        ggplot2::theme(legend.position = "bottom", legend.key = element_rect(colour = "white")) + 
        ggplot2::labs(color = "")
    return(plot)
}

